master/internal/kubernetes/pod.go

```diff
@@ -101,9 +153,191 @@ func (p *pod) startPod(ctx *actor.Context) error {
+				ctx.Tell(p.taskHandler, rsc)
+			}
+
+			ctx.Log().WithField("pod", p.podName).Infof(
+				"transitioning pod state from %s to %s",
+				p.container.State, containerState)
+			p.container = p.container.Transition(containerState)
+
+			rsc := sproto.ContainerStateChanged{Container: p.container}
+			ctx.Tell(p.taskHandler, rsc)
+		}
+
+	case k8sV1.PodRunning:
+		if p.container.State != container.Running {
+			p.container = p.container.Transition(container.Running)
+
+			logStreamer, err := newPodLogStreamer(p.podInterface, p.podName, ctx.Self())
+			if err != nil {
+				return err
+			}
+			var ok bool
+			p.logStreamer, ok = ctx.ActorOf(fmt.Sprintf("%s-logs", p.podName), logStreamer)
+			if !ok {
+				return errors.Errorf("log streamer already exists")
+			}
+
+			ctx.Tell(p.logStreamer, streamLogs{})
+			ctx.Tell(p.taskHandler, sproto.ContainerStateChanged{Container: p.container})
+			ctx.Tell(p.cluster, sproto.PodStarted{
+				ContainerID: p.container.ID,
+				IP:          p.pod.Status.PodIP,
+				Ports:       p.ports,
+			})
+		}
+
+	case k8sV1.PodFailed:
+		if p.container.State != container.Terminated {
+			p.container = p.container.Transition(container.Terminated)
+
+			exitCode, exitMessage, err := getExitCodeAndMessage(p.pod)
+			if err != nil {
+				return err
+			}
+			ctx.Log().WithField("pod", p.podName).Infof("pod failed: %d %s", exitCode, exitMessage)
+
+			exitCodeConverted := agent.ExitCode(exitCode)
+			containerStopped := agent.ContainerStopped{
+				Failure: &agent.ContainerFailure{
+					FailureType: agent.ContainerFailed,
+					ErrMsg:      exitMessage,
+					ExitCode:    &exitCodeConverted,
+				},
+			}
+
+			ctx.Tell(p.taskHandler, sproto.ContainerStateChanged{
+				Container:        p.container,
+				ContainerStopped: &containerStopped,
+			})
+
+			ctx.Tell(p.cluster, sproto.PodTerminated{
+				ContainerID:      p.container.ID,
+				ContainerStopped: &containerStopped,
+			})
+
+			ctx.Self().Stop()
+		}
+
+	case k8sV1.PodSucceeded:
+		if p.container.State != container.Terminated {
+			p.container = p.container.Transition(container.Terminated)
+
+			ctx.Log().WithField("pod", p.podName).Infof("pod exited successfully")
+			containerStopped := agent.ContainerStopped{}
+
+			ctx.Tell(p.taskHandler, sproto.ContainerStateChanged{
+				Container:        p.container,
+				ContainerStopped: &containerStopped,
+			})
+
+			ctx.Tell(p.cluster, sproto.PodTerminated{
+				ContainerID:      p.container.ID,
+				ContainerStopped: &containerStopped,
+			})
+
+			ctx.Self().Stop()
+		}
+	default:
+		return errors.Errorf(
+			"unexpected pod status %s for pod %s", msg.updatedPod.Status.Phase, p.podName)
+	}
+
+	return nil
+}
+
+func (p *pod) deleteKubernetesResources(ctx *actor.Context) error {
+	if p.resourcesDeleted {
+		return nil
+	}
+
+	ctx.Log().WithField("pod", p.podName).Infof("deleting pod")
+	var gracePeriod int64 = 1
+	err := p.podInterface.Delete(p.podName, &metaV1.DeleteOptions{GracePeriodSeconds: &gracePeriod})
```

Comment:
question: Just to check, is this an internal Kubernetes bookkeeping operation or is it waiting for a container to update? One second seems a bit short if the latter.

Reply:
From my understanding this graceful period is how long k8s waits to send a SIGKILL after sending as SIGTERM to the pods. Do you think we should increase this period? I haven't explored how consistently SIGTERM is able to get our pods to exit out.

Reply:
Yeah, I'm also totally guessing, but a bit longer feels reasonable; we probably don't want to be slinging SIGKILLs around without reason as a general rule. Something between 10 and 30 seconds, maybe.

Reply:
Ok bumped up to 15 seconds. If we see weird behavior from this, can always revisit in the future.